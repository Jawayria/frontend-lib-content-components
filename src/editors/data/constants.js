// maybe put the string mapping in here?