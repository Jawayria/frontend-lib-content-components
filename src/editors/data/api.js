import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
import { ActionStates, normalizeContent } from './constants';

async function getAsync(setValue, setError, setLoading, params) {
  try {
    setLoading(ActionStates.IN_PROGRESS);
    const result = await getAuthenticatedHttpClient().get(...params);
    setValue(result);
  } catch (e) {
    setError(e);
  } finally {
    setLoading(ActionStates.FINISHED);
  }
}

async function saveAsync(setInProgress, setResponse, params) {
  try {
    const result = await getAuthenticatedHttpClient().post(...params);
    setResponse(result);
  } catch (e) {
    setResponse(e);
  } finally {
    setInProgress(ActionStates.FINISHED);
  }
}

export function fetchBlockById(setValue, setError, setLoading, blockId, studioEndpointUrl) {
  const url = `${studioEndpointUrl}/xblock/${blockId}`;
  getAsync(setValue, setError, setLoading, [url]);
}

export function fetchUnitById(setValue, setError, setLoading, blockId, studioEndpointUrl) {
  const url = `${studioEndpointUrl}/xblock/${blockId}?fields=ancestorInfo`;
  getAsync(setValue, setError, setLoading, [url]);
}

export function saveBlock(blockId, blockType, courseId, studioEndpointUrl, content, setInProgress, setResponse) {
  const normalizedContent = normalizeContent(blockType, content, blockId, courseId);
  const url = `${studioEndpointUrl}/xblock/${encodeURI(blockId)}`;
  const params = [url, normalizedContent];
  saveAsync(setInProgress, setResponse, params);
}
