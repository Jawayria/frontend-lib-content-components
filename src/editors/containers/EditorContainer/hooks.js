import { useSelector } from 'react-redux';

import { RequestKeys } from '../../data/constants/requests';
import { selectors } from '../../data/redux';
import * as appHooks from '../../hooks';

export const {
  navigateCallback,
  nullMethod,
  saveBlock,
} = appHooks;

export const handleSaveClicked = ({ getContent, dispatch }) => (
  () => saveBlock({ content: getContent(), dispatch })
);
export const handleCancelClicked = ({ onClose }) => {
  if (onClose) {
    return onClose;
  }
  return navigateCallback(useSelector(selectors.app.returnUrl));
};
export const isInitialized = () => useSelector(selectors.app.isInitialized);
export const saveFailed = () => useSelector((state) => (
  selectors.requests.isFailed(state, { requestKey: RequestKeys.saveBlock })
));
