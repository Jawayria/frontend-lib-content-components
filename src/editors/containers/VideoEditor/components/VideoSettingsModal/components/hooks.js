import { useSelector } from 'react-redux';

import { keyStore } from '../../../../../utils';
import { actions, selectors } from '../../../../../data/redux';

export const selectorKeys = keyStore(selectors.video);

export const widgetValue = (key, dispatch) => ({
  formValue: useSelector(selectors.video[key]),
  setFormValue: (val) => dispatch(actions.video.load({ [key]: val })),
});

export default {
  selectorKeys,
  widgetValue,
};
