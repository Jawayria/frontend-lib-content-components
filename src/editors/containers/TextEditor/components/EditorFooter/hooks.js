import { useSelector } from 'react-redux';

import { RequestKeys } from '../../../../data/constants/requests';
import { selectors, thunkActions } from '../../../../data/redux';
import * as textEditorHooks from '../../hooks';

export const {
  navigateCallback,
  nullMethod,
  saveBlock,
} = textEditorHooks;

export const handleSaveClicked = ({ dispatch, editorRef }) => () => saveBlock({
  editorRef,
  saveBlockContent: (...args) => dispatch(thunkActions.app.saveBlock(...args)),
  returnUrl: useSelector(selectors.app.returnUrl),
});

export const handleCancelClicked = () => navigateCallback(useSelector(selectors.app.returnUrl));

export const isInitialized = () => useSelector(selectors.app.isInitialized);

export const saveFailed = () => useSelector((state) => (
  selectors.requests.isFailed(state, { requestKey: RequestKeys.saveBlock })
));

export const studioEndpointUrl = () => useSelector(selectors.app.studioEndpointUrl);
