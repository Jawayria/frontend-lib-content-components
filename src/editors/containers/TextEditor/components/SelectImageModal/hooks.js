import React from 'react';
import { useDispatch } from 'react-redux';

import { thunkActions } from '../../../../data/redux';
import * as module from './hooks';
import { sortFunctions, sortKeys } from './utils';

export const state = {
  images: (val) => React.useState(val),
  highlighted: (val) => React.useState(val),
  searchString: (val) => React.useState(val),
  sortBy: (val) => React.useState(val),
};

export const searchAndSortHooks = () => {
  const [searchString, setSearchString] = module.state.searchString('');
  const [sortBy, setSortBy] = module.state.sortBy(sortKeys.dateNewest);
  return {
    searchString,
    onSearchChange: e => setSearchString(e.target.value),
    clearSearchString: () => setSearchString(''),
    sortBy,
    onSortClick: key => () => setSortBy(key),
  };
};

export const filteredList = ({ searchString, imageList }) => imageList.filter(
  ({ displayName }) => displayName.toLowerCase().includes(searchString.toLowerCase()),
);

export const displayList = ({ sortBy, searchString, images }) => module.filteredList({
  searchString,
  imageList: Object.values(images),
}).sort(sortFunctions[sortBy in sortKeys ? sortKeys[sortBy] : sortKeys.dateNewest]);

export const imgListHooks = ({
  setSelection,
  searchSortProps,
}) => {
  const dispatch = useDispatch();
  const [images, setImages] = module.state.images({});
  const [highlighted, setHighlighted] = module.state.highlighted(null);

  React.useEffect(() => {
    dispatch(thunkActions.app.fetchImages({ setImages }));
  }, []);

  return {
    images,
    // highlight by id
    selectBtnProps: {
      disabled: !highlighted,
      onClick: () => setSelection(images[highlighted]),
    },
    galleryProps: {
      highlighted,
      onHighlightChange: e => setHighlighted(e.target.value),
      displayList: module.displayList({ ...searchSortProps, images }),
    },
  };
};

export const fileInputHooks = ({ setSelection }) => {
  const dispatch = useDispatch();
  const ref = React.useRef();
  const click = () => ref.current.click();
  const addFile = (e) => {
    dispatch(thunkActions.app.uploadImage({
      file: e.target.files[0],
      setSelection,
    }));
  };

  return {
    click,
    addFile,
    ref,
  };
};

export const imgHooks = ({ setSelection }) => {
  const searchSortProps = module.searchAndSortHooks();
  const imgList = module.imgListHooks({ setSelection, searchSortProps });
  const fileInput = module.fileInputHooks({ setSelection });
  const { selectBtnProps, galleryProps } = imgList;

  return {
    fileInput,
    galleryProps,
    selectBtnProps,
    searchSortProps,
  };
};

export default {
  imgHooks,
};
