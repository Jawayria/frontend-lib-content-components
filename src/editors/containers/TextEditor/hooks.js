import { useState } from 'react';
import * as module from './hooks';

export const addImageUploadButton = (openModal, setImage) => (editor) => {
  editor.ui.registry.addButton('imageuploadbutton', {
    icon: 'image',
    onAction: openModal,
  });
  editor.ui.registry.addButton('editimagesettings', {
    icon: 'image',
    onAction() {
      const imgHTML = editor.selection.getNode();
      setImage({
        externalUrl: imgHTML.src,
        altText: imgHTML.alt,
        width: imgHTML.width,
        height: imgHTML.height,
      });
      openModal();
    },
  });
};

export const initializeEditorRef = (setRef, initializeEditor) => (editor) => {
  setRef(editor);
  initializeEditor();
};

// for toast onClose to avoid console warnings
export const nullMethod = () => {};

export const editorConfig = ({
  setEditorRef,
  blockValue,
  openModal,
  initializeEditor,
  setSelection,
}) => ({
  onInit: (evt, editor) => module.initializeEditorRef(setEditorRef, initializeEditor)(editor),
  initialValue: blockValue ? blockValue.data.data : '',
  init: {
    setup: module.addImageUploadButton(openModal, setSelection),
    plugins: 'link codesample emoticons table charmap code autoresize image imagetools',
    menubar: false,
    toolbar: 'undo redo | formatselect | '
      + 'bold italic backcolor | alignleft aligncenter '
      + 'alignright alignjustify | bullist numlist outdent indent |'
      + 'imageuploadbutton | link | emoticons | table | codesample | charmap |'
      + 'removeformat | hr |code',
    imagetools_toolbar: 'rotateleft rotateright | flipv fliph | editimagesettings',
    imagetools_cors_hosts: ['courses.edx.org'],
    height: '100%',
    content_style: 'body { font-family:Helvetica,Arial,sans-serif; font-size:14px }',
    min_height: 1000,
    branding: false,
  },
});

export const modalToggle = () => {
  const [isOpen, setIsOpen] = useState(false);
  return {
    isOpen,
    openModal: () => setIsOpen(true),
    closeModal: () => setIsOpen(false),
  };
};
