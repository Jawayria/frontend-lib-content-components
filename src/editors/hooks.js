import { useEffect } from 'react';

import * as module from './hooks';

export const initializeApp = ({ initialize, data }) => useEffect(
  () => initialize(data),
  [],
);

export const navigateTo = (destination) => {
  window.location.assign(destination);
};

export const navigateCallback = (destination) => () => module.navigateTo(destination);
