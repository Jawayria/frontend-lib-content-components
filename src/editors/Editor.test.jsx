import React from 'react';
import { shallow } from 'enzyme';
import { Editor, mapDispatchToProps } from './Editor';
import { thunkActions } from './data/redux';
import * as hooks from './hooks';

const prepareEditorRefSpy = jest.spyOn(hooks, 'prepareEditorRef');

const props = {
  courseId: 'course-v1:edX+DemoX+Demo_Course',
  blockType: 'html',
  blockId: 'block-v1:edX+DemoX+Demo_Course+type@html+block@030e35c4756a4ddc8d40b95fbbfff4d4',
  studioEndpointUrl: 'fakeurl.com',
  initialize: jest.fn(),
};

jest.mock('./data/redux', () => ({
  actions: {
    app: {
      initializeEditor: jest.fn().mockName('actions.app.initializeEditor'),
    },
  },
  thunkActions: {
    app: {
      initialize: jest.fn().mockName('thunkActions.app.initialize'),
    },
  },
  selectors: {
    app: {
      blockValue: jest.fn(state => ({ blockValue: state })),
    },
    requests: {
      isFailed: jest.fn((state, params) => ({ isFailed: { state, params } })),
      isFinished: jest.fn((state, params) => ({ isFailed: { state, params } })),
    },
  },
}));

describe('Editor', () => {
  describe('snapshots', () => {
    test('renders no editor when ref isnt ready', () => {
      prepareEditorRefSpy.mockImplementationOnce(
        () => ({ editorRef: null, refReady: false, setEditorRef: jest.fn() }),
      );
      expect(shallow(<Editor {...props} />)).toMatchSnapshot();
    });
    test('renders editor when ref is ready', () => {
      prepareEditorRefSpy.mockImplementationOnce(
        () => ({ editorRef: { current: 'ref' }, refReady: true, setEditorRef: jest.fn().mockName('setEditorRef') }),
      );
      expect(shallow(<Editor {...props} />)).toMatchSnapshot();
    });
    test('presents error message if no relevant editor found and ref ready', () => {
      prepareEditorRefSpy.mockImplementationOnce(
        () => ({ editorRef: { current: 'ref' }, refReady: true, setEditorRef: jest.fn().mockName('setEditorRef') }),
      );
      expect(shallow(<Editor
        {...props}
        blockType="fAkEBlock"
      />)).toMatchSnapshot();
    });
  });
  describe('mapDispatchToProps', () => {
    test('initialize from thunkActions.app.initialize', () => {
      expect(mapDispatchToProps.initialize).toEqual(thunkActions.app.initialize);
    });
  });
});
