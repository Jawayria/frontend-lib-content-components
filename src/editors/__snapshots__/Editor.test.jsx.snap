// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor snapshots presents error message if no relevant editor found and ref ready 1`] = `
<div
  className="d-flex flex-column vh-100"
>
  <div
    aria-label="fAkEBlock"
    className="pgn__modal-fullscreen"
    role="dialog"
  >
    <EditorHeader
      editorRef={
        Object {
          "current": "ref",
        }
      }
    />
    <FormattedMessage
      defaultMessage="Error: Could Not find Editor"
      description="Error Message Dispayed When An unsopported Editor is desired in V2"
      id="authoring.editorpage.selecteditor.error"
    />
    <EditorFooter
      editorRef={
        Object {
          "current": "ref",
        }
      }
    />
  </div>
</div>
`;

exports[`Editor snapshots renders "html" editor when ref is ready 1`] = `
<div
  className="d-flex flex-column vh-100"
>
  <div
    aria-label="html"
    className="pgn__modal-fullscreen"
    role="dialog"
  >
    <EditorHeader
      editorRef={
        Object {
          "current": "ref",
        }
      }
    />
    <TextEditor
      setEditorRef={[MockFunction setEditorRef]}
    />
    <EditorFooter
      editorRef={
        Object {
          "current": "ref",
        }
      }
    />
  </div>
</div>
`;

exports[`Editor snapshots renders "problem" editor when ref is ready 1`] = `
<div
  className="d-flex flex-column vh-100"
>
  <div
    aria-label="problem"
    className="pgn__modal-fullscreen"
    role="dialog"
  >
    <EditorHeader
      editorRef={
        Object {
          "current": "ref",
        }
      }
    />
    <ProblemEditor
      setEditorRef={[MockFunction setEditorRef]}
    />
    <EditorFooter
      editorRef={
        Object {
          "current": "ref",
        }
      }
    />
  </div>
</div>
`;

exports[`Editor snapshots renders "video" editor when ref is ready 1`] = `
<div
  className="d-flex flex-column vh-100"
>
  <div
    aria-label="video"
    className="pgn__modal-fullscreen"
    role="dialog"
  >
    <EditorHeader
      editorRef={
        Object {
          "current": "ref",
        }
      }
    />
    <VideoEditor
      setEditorRef={[MockFunction setEditorRef]}
    />
    <EditorFooter
      editorRef={
        Object {
          "current": "ref",
        }
      }
    />
  </div>
</div>
`;

exports[`Editor snapshots renders no editor when ref isnt ready 1`] = `
<div
  className="d-flex flex-column vh-100"
>
  <div
    aria-label="AblOck"
    className="pgn__modal-fullscreen"
    role="dialog"
  />
</div>
`;
